src/cfg/File.h

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (C) 1996-2021 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#ifndef _SQUID__SRC_CFG_FILE_H
+#define _SQUID__SRC_CFG_FILE_H
+
+#include "base/Here.h"
+#include "cfg/forward.h"
+#include "sbuf/SBuf.h"
+
+#include <cstdio>
+#include <string>
+
+namespace Cfg
+{
+
+/**
+ * Class used to store required information for the current
+ * configuration file.
+ */
+class File
+{
+public:
+    File(const char *path) : filePath(path) {}
+    ~File();
+
+    /// \return true if the configuration file is open
+    bool isOpen() const { return bool(fd); }
+
+    /// \return the next line to be parsed from this file
+    SBuf nextLine();
+
+    /// \return the configuration file name and line number being processed
+    SourceLocation lineInfo() const { return SourceLocation("parsing", filePath.c_str(), lineNo); }
+
+    /// open and load contents from the file
+    void tryLoadFile();
+
+private:
+    std::string filePath; ///< The file path
+    FILE *fd = nullptr; ///< Pointer to the file.
+    SBufList fileData; ///< bytes read from this file in advance of parsing
```

Comment:
We need to document the list order and the contents of a list element (because it is not a "byte" as the current description might imply).

We should also document why bytes are stored in a _list_ because that is far from obvious.

Where possible, we should also avoid predicting how we are going to be used by others.

Here is one possible fix but please adjust as needed:

```suggestion

    /// Random-size blocks of raw configuration bytes, in read(2) order.
    /// We do not concatenate these blocks to avoid overflowing SBuf.
    SBufList fileData;

```

Reply:
> (because it is not a "byte" as the current description might imply).

What do you think it is then?
fread(2) documentation says it fills the given buffer with bytes.

Reply:
> Alex: We need to document ... the contents of a list element (because it is not a "byte" as the current description might imply).

> Amos: What do you think it is then?

The answer to that question depends on the code and intent. There were too many problems in the associated code for me to be certain what fileData is meant to be and what it actually is. However, I had enough information to know that the description needed work, as detailed in my change request. I hope _you_ will do that work. fileData is a list of X. Please adjust description to document X storage order and what _you think_ X is. The fact that fileData is a list is obvious because of its SBufList type. The fact that X storage is SBuf is similarly obvious. It is not obvious what X is, but X is not a "byte", in any practical interpretation of that word, even if some X values may store a single byte read from the file. X could be a raw line as read from the file or a random-size block of raw data read from the file or something else instead. You know what it should be. And please do not forget about documenting the order of X.

Please do not misinterpret the above discussion as an implication that fileData type itself is correct (or incorrect). That is a separate question, but it would be much easier to answer that question correctly once we know what the current fileData is.

Reply:
> You are looking at the lowest level of code in a literal sense

I do not.

> My documentation is filling out details of the abstraction unstated in the variable name.

A container of ordered items should make the order and meaning of its items clear. `fileData` failed to do that. Arguing about these basics is a waste of time.